### ISSUE TITLE
Emit constants for string literal enum types

### ISSUE THREAD

**Issue by tonisuiboupin** [raphaeljolivet/java2typescript#68]
@andyperlitch implemented string literal types, but it would be useful to also emit constants for string literal enum type values

**Comment by tonisuiboupin** [comment]
I'll sent PR soon

**Comment by andyperlitch** [comment]
What is the utility of this change? IDEs should be fine autocompleting/type-checking for string literal types. This just adds more definitions unnecessarily. At the very least, you should have implemented this as a new writer, not assumed people wanting string literal types would also want this.

**Comment by atsu85** [comment]
@andyperlitch, i thought about this feature myself right after merging Your PR, because when You want to check if value is equal to specific constant, You couldn't reference it previously. So You'd do smth like:
```TypeScript
if(this.type === "VAL1") ...
```
and when enum constant "VAL1" is renamed/removed, then the TypeScript code still compiles without any problems, even tough the if statement will never evaluate to true.
So it makes sense to have the constants generated as well, so You'd be able to reference them
```TypeScript
if(this.type === MyEnumValues.VAL1) ...
```
and get a compile-time error when "VAL1" is renamed/removed.

**Comment by atsu85** [comment]
@andyperlitch 
> At the very least, you should have implemented this as a new writer, ...

I wouldn't implement it as a new writer - it seems to me that this class is the best place for it right now.

> ... not assumed people wanting string literal types would also want this.

I thought about it briefly - i guess there are two cases when You might want to avoid generating the constants in addition to types:
1) When You have two enums with tha same prefiks where other has suffix "Values", for example MyEnum and MyEnumValues. In this case EnumTypeToStringLiteralTypeWriter would like to write:
* type MyEnum
* class MyEnumValues
* type MyEnumValues // conflicts with class containing values of MyEnum type
* class MyEnumValuesValues
2) When You have lots of enums and You want to cut down the bundle size.

If you want to avoid generating the values, I'd gladly accept the PR that makes this feature configurable, but i think it is best to leave it on as default.

**Comment by andyperlitch** [comment]
>and when enum constant "VAL1" is renamed/removed, then the TypeScript code still compiles without any problems, even tough the if statement will never evaluate to true.

Actually this is not true (not sure what IDE/typescript version you use). Take a look at my editor in this situation (using VSCode):

![image](https://cloud.githubusercontent.com/assets/1390651/22712972/846c5544-ed3b-11e6-92af-ee7ba65034c7.png)

As far as I can see, that was the biggest argument. And as you say, this change increases the chances for inadvertent collision.

I would still opt for simply rolling back this change, but in lieu of that I will open a PR.

**Comment by atsu85** [comment]
> (not sure what IDE/typescript version you use). Take a look at my editor in this situation (using VSCode):

hmm, that would be good news, if it wasn't just a magic from VSCode. 

I experimented with TypeScript 1.8.10 using Gulp and IntelliJ Idea - neither produced an error when wrong enum constant was used.

> As far as I can see, that was the biggest argument.

Correct.  I hope that newer TS compiler emits the error, not VSCode itself - if so, it makes sense to revert this change (or at least change the default of emitting types to false, if someone will find it useful and is stuck on version that doesn't emit the error). 

> I will open a PR

Good, right now i don't have the time to experiment more with it myself.

**Comment by andyperlitch** [comment]
>if it wasn't just a magic from VSCode.
VSCode is good, but its not that good. Type-checking for this actually *IS* available, all the way back to v1.8.10. I just ran a quick test:

```bash
$ cat test.ts
type Foo = 'foo' | 'bar';
const f: Foo = 'baz';
$ ./node_modules/.bin/tsc -v
Version 1.8.10
$ ./node_modules/.bin/tsc test.ts
test.ts(2,7): error TS2322: Type '"baz"' is not assignable to type '"foo" | "bar"'.
  Type '"baz"' is not assignable to type '"bar"'.
```

Perhaps you have some config option to silence this error? At any rate, given this test I strongly suggest we simply roll this change back.

**Comment by atsu85** [comment]
Yeah,
> `type Foo = 'foo' | 'bar';`
`const f: Foo = 'baz';`

That wasn't what i was talking about, glad You came to the same conclusion :)

>  from typescript v2.0.2 and onward, you do get an error:

That is what i hoped :)

> For backwards-compatibility with pre-2.0.2 typescript, I suppose this change is okay, but it should be opt-in.

Exactly :)
